apm-agent-api/src/main/java/co/elastic/apm/api/ElasticApm.java

```diff
@@ -1,40 +1,121 @@
 package co.elastic.apm.api;
 
+/**
+ * This class can be used to statically access the {@link Tracer}.
+ * <p>
+ * You can store the Tracer as an instance like so:
+ * </p>
+ * <pre>{@code
+ * private static final Tracer tracer = ElasticApm.get();
+ * }</pre>
+ * <p>
+ * Then you access the tracer to set a custom transaction name,
+ * for example:
+ * <pre>{@code
+ * tracer.currentTransaction().setName("SuchController#muchMethod");
+ * }</pre>
+ * <p/>
+ */
 public class ElasticApm implements Tracer {
 
+    /**
+     * The singleton instance of the Tracer.
+     */
     private static final ElasticApm INSTANCE = new ElasticApm();
 
     private static Tracer tracer = NoopTracer.INSTANCE;
 
+    /**
+     * {@link ElasticApm} may only be accessed via {@link #get()}
+     */
+    private ElasticApm() {
+    }
+
+    /**
+     * Returns the tracer implementation.
+     *
+     * @return the tracer implementation (never <code>null</code>)
+     */
     public static Tracer get() {
         return INSTANCE;
     }
 
-    public static void register(Tracer tracer) {
-        ElasticApm.tracer = tracer;
+    /**
+     * Statically registers the tracer implementation,
+     * so that it can be accessed via {@link ElasticApm#get()}
+     * <p>
+     * This method is called by the actual {@link Tracer} implementation.
+     * </p>
+     * <p>
+     * Users are not supposed to set a custom instrumentation so this method must not be <code>public</code>.
+     * Otherwise it would be part of the public API.
+     * </p>
+     * <p>
+     * Loading the concrete implementation from here is also not a good option,
+     * as it could not be easily set to a different (possibly mocked) implementation in tests.
+     * </p>
+     * <p>
+     * So the solution is to provide a package-private method which the implementation can call.
+     * What's not so nice about this solution is that the implementation has to have a class in
+     * the same package as the api in order to be able to call this package private method.
+     * But as this is not a public method,
+     * the concrete registration mechanism can always be changed later without breaking compatibility.
+     * </p>
+     *
+     * @param tracer The tracer implementation to register.
+     */
+    static void register(Tracer tracer) {
```

Comment:
Why is this method static? You are inside a singleton, it should be safe not to have this static which in return will make this less testable.

Reply:
Because the `tracer` variable is static atm. But we could definitely change that to being non static. I think in this case, it does not really affect testability, but it's always a good thing to minimize the usage of the `static` keyword.